src/imbalances_script.py

```diff
@@ -0,0 +1,255 @@
+    """
+    for chain_name, url in CHAIN_RPC_ENDPOINTS.items():
+        web3 = Web3(Web3.HTTPProvider(url))
+        if not web3.is_connected():
+            print(f"Could not connect to {chain_name}.")
+            continue
+        try:
+            web3.eth.get_transaction_receipt(tx_hash)
+            print(f"Transaction found on {chain_name}.")
+            return chain_name, web3
+        except Exception as e:
+            print(f"Transaction not found on {chain_name}: {e}")
+    raise ValueError(f"Transaction hash {tx_hash} not found on any chain.")
+
+def _to_int(value: str | int) -> int:
+    """Convert hex string or integer to integer."""
+    try:
+        return int(value, 16) if isinstance(value, str) and value.startswith('0x') else int(value)
+    except ValueError:
+        print(f"Error converting value {value} to integer.")
+
+class RawTokenImbalances:
+    def __init__(self, web3: Web3, chain_name: str):
+        self.web3 = web3
+        self.chain_name = chain_name
+
+    def get_transaction_receipt(self, tx_hash: str) -> Optional[Dict]:
+        """
+        Get the transaction receipt from the provided web3 instance.
+        """
+        try:
+            return self.web3.eth.get_transaction_receipt(tx_hash)
+        except Exception as e:
+            print(f"Error getting transaction receipt: {e}")
+            return None
+
+    def get_transaction_trace(self, tx_hash: str) -> Optional[List[Dict]]:
+        """ Function used for retreiving trace to identify ETH transfers. """
+        try:
+            res = self.web3.tracing.trace_transaction(tx_hash)
+            return res
+        except Exception as err:
+            print(f"Error occurred while fetching transaction trace: {err}")
+            return None
+
+    def extract_actions(self, traces: List[AttributeDict], address: str) -> List[Dict]:
+        """ Identify transfer events in trace involving the specified contract. """
+        normalized_address = Web3.to_checksum_address(address)
+        actions = []
+        # input_field = '0x' denotes a native ETH transfer event, which we want to filter for
+        input_field: str = '0x'
+        for trace in traces:
+            if isinstance(trace, AttributeDict):
+                action = trace.get('action', {})
+                input_value = action.get('input', b"").hex()
+                # filter out action if involved in an ETH transfer event
+                if input_value == input_field and (
+                    Web3.to_checksum_address(action.get('from', '')) == normalized_address or
+                    Web3.to_checksum_address(action.get('to', '')) == normalized_address
+                ):
+                    actions.append(dict(action))
+        return actions
+
+    def calculate_native_eth_imbalance(self, actions: List[Dict], address: str) -> int:
+        """Extract ETH imbalance from transfer actions."""
+        # inflow is the total value transferred to CoW contract
```

Comment:
Function works for any address, as far as i understand, and not just for the CoW contract.